### DISCUSSION TITLE
make plugin use in browser, instead of typescript?

### DISCUSSION THREAD

**Discussion by boroco** [idiotWu/smooth-scrollbar#455] [Q&A] [answered]
Hi Team,
https://github.com/idiotWu/smooth-scrollbar/issues/198
https://github.com/idiotWu/smooth-scrollbar/issues/408
I need smooth and slow scrolling to the top or bottom of the page.

i need an effect like this issue, but i can't write them in javascript browser. It always says " Cannot read properties of undefined (reading 'pluginName')". Can you help me with the code using javascript?, instead of typescript. I think a lot of people need this.

Thank you author.

**Comment by dolphin-wood** [comment]
You can check https://github.com/idiotWu/smooth-scrollbar/issues/104#issuecomment-335286699 to see how to write a plugin in es5 syntax.

P.S. are you the one who sent me the email about the same question?

**Reply by boroco** [reply]
@idiotWu 
That's also the question my friend asked in the email, we all have the same question.

**Comment by dolphin-wood** [comment, marked as answer]
To be more specific:

```js
var Scrollbar = window.Scrollbar;

// see also: https://github.com/idiotWu/smooth-scrollbar/issues/198#issuecomment-496264237
function EdgeEasingPlugin() {
  Scrollbar.ScrollbarPlugin.apply(this, arguments);

  this._remainMomentum = {
    x: 0,
    y: 0,
  };
}

EdgeEasingPlugin.prototype = Object.create(Scrollbar.ScrollbarPlugin.prototype);

EdgeEasingPlugin.prototype.transformDelta = function(delta) {
  const {
    limit,
    offset,
  } = this.scrollbar;

  const x = this._remainMomentum.x + delta.x;
  const y = this._remainMomentum.y + delta.y;

  // clamps momentum within [-offset, limit - offset]
  this.scrollbar.setMomentum(
    Math.max(-offset.x, Math.min(x, limit.x - offset.x)),
    Math.max(-offset.y, Math.min(y, limit.y - offset.y)),
  );

  return { x: 0, y: 0 };
}

EdgeEasingPlugin.prototype.onRender = function onRender(remainMomentum) {
  Object.assign(this._remainMomentum, remainMomentum);
}

EdgeEasingPlugin.pluginName = 'edgeEasing';

```

**Comment by boroco** [comment]
@idiotWu 
I copy your code, then call Scrollbar.use(EdgeEasingPlugin); doesn't seem to work, it shows error "Uncaught ReferenceError: EdgeEasingPlugin is not defined".
Can you send me a little more detail?

**Reply by dolphin-wood** [reply]
Did you define and use the plugin in the same scope?

**Reply by boroco** [reply]
![image](https://user-images.githubusercontent.com/100500240/155867128-f3db82f0-a39e-4da1-b2d9-5658c1dfd376.png)
@idiotWu 
it shows this error.

**Reply by boroco** [reply]
and Uncaught ReferenceError: EdgeEasingPlugin is not defined

**Reply by dolphin-wood** [reply]
So your build system does not support the [trailing commas](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Trailing_commas). Remove the last comma at line 785 should make it work.

**Reply by boroco** [reply]
Uncaught TypeError: Cannot read properties of undefined (reading 'ScrollbarPlugin')
![image](https://user-images.githubusercontent.com/100500240/155867232-d4ade598-eb2b-486b-a107-965bdcf6cb1d.png)
@idiotWu

**Reply by dolphin-wood** [reply]
What's the name of your `Scrollbar` class?

**Reply by boroco** [reply]
Thanks @idiotWu 
i changed
var Scrollbar = window.Scrollbar; to import Scrollbar from 'smooth-scrollbar';
it's working now.

**Comment by boroco** [comment]
Oh, Perfect.
Source code i use:
```
import Scrollbar from 'smooth-scrollbar';
/* or var Scrollbar = window.Scrollbar; */

`/ see also: https://github.com/idiotWu/smooth-scrollbar/issues/198#issuecomment-496264237
  function EdgeEasingPlugin() {
    Scrollbar.ScrollbarPlugin.apply(this, arguments);

    this._remainMomentum = {
      x: 0,
      y: 0,
    };
  }

  EdgeEasingPlugin.prototype = Object.create(Scrollbar.ScrollbarPlugin.prototype);

  EdgeEasingPlugin.prototype.transformDelta = function(delta) {
    const {
      limit,
      offset,
    } = this.scrollbar;

    const x = this._remainMomentum.x + delta.x;
    const y = this._remainMomentum.y + delta.y;

    // clamps momentum within [-offset, limit - offset]
    this.scrollbar.setMomentum(
      Math.max(-offset.x, Math.min(x, limit.x - offset.x)),
      Math.max(-offset.y, Math.min(y, limit.y - offset.y))
    );

    return { x: 0, y: 0 };
  }

  EdgeEasingPlugin.prototype.onRender = function onRender(remainMomentum) {
    Object.assign(this._remainMomentum, remainMomentum);
  }

  EdgeEasingPlugin.pluginName = 'edgeEasing';
Scrollbar.use(EdgeEasingPlugin);`
```

**Comment by boroco** [comment]
@idiotWu 
I see a lot of people need scroll animation like this, if it is added in v9 that would be great. the user will have a lot of options when scrolling.
  Thanks the author.
